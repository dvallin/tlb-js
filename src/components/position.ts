import { Vector } from '../spatial'

export interface PositionComponent {
  level: number
  position: Vector
}

export function centeredPosition(level: number, position: Vector): PositionComponent {
  return { level, position: position.center }
}
