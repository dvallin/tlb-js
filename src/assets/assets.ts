import { Asset, Cover } from '../components/asset'
import { Vector } from '../spatial'
import { features, generators } from './features'
import { Feature } from '../components/feature'
import { DialogType } from './dialogs'

const s1x1 = new Vector([1, 1])
const s1x2 = new Vector([1, 2])
const s3x1 = new Vector([3, 1])
const s2x2 = new Vector([2, 2])
function asset(name: string, size: Vector, cover: Cover, dialog: DialogType | undefined, feature: (index: number) => Feature): Asset {
  return { name, size, cover, hasInventory: true, dialog, feature }
}

const assetsDefinition = {
  door: asset('metal door', s3x1, 'full', undefined, () => features['door']),
  locker: asset('a locker', s1x1, 'full', undefined, () => features['locker']),
  trash: asset('some trash', s1x1, 'none', undefined, () => features['trash']),
  loot: asset('a dead body', s1x1, 'none', undefined, () => features['loot']),
  table: asset('a table', s1x2, 'partial', undefined, () => features['table']),
  generator: asset('a generator', s2x2, 'full', undefined, generators['block']),
  elevator: asset('an elevator', s1x1, 'none', 'elevator', generators['elevator']),
}

export type AssetType = keyof typeof assetsDefinition
export const assets: { [key in AssetType]: Asset } = assetsDefinition
