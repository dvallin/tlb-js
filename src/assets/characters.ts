import { BodyPart, CharacterStats, CharacterStatsComponent, createCharacterStatsComponent } from '../components/character-stats'
import { TlbWorld } from '../tlb'
import { features, FeatureType } from './features'
import { FeatureComponent } from '../components/feature'
import { FovComponent } from '../components/fov'
import { HasActionComponent } from '../components/action'
import { ItemComponent, InventoryComponent, EquipedItemsComponent } from '../components/items'
import { ActiveEffectsComponent } from '../components/effects'
import { CharacterCreator } from '../generative/complex-embedder'
import { Entity } from '../ecs/entity'
import { AiComponent } from '../components/ai'
import { ActionType } from './actions'
import { ItemType } from './items'
import { addDialog } from './dialogs'
import { TriggersComponent } from '../components/trigger'

function humanoidBodyParts(core: number, head: number, limp: number): { [key: string]: BodyPart } {
  return {
    head: { type: 'head', itemAttachments: ['helmet'], health: head },
    leftArm: { type: 'arm', itemAttachments: ['glove', 'weapon'], health: limp },
    rightArm: { type: 'arm', itemAttachments: ['glove', 'weapon'], health: limp },
    torso: { type: 'torso', itemAttachments: ['armor'], health: core },
    leftLeg: { type: 'leg', itemAttachments: ['pants', 'boots'], health: limp },
    rightLeg: { type: 'leg', itemAttachments: ['pants', 'boots'], health: limp },
  }
}

const charactersStatsDefinition = {
  player: {
    bodyParts: humanoidBodyParts(14, 5, 5),
    strength: 5,
    movement: 4,
    actions: 4,
    aim: 10,
  },
  guard: {
    bodyParts: humanoidBodyParts(3, 1, 1),
    strength: 5,
    movement: 3,
    actions: 3,
    aim: 6,
  },
  eliteGuard: {
    bodyParts: humanoidBodyParts(5, 3, 3),
    strength: 8,
    movement: 2,
    actions: 5,
    aim: 6,
  },
  boss: {
    bodyParts: humanoidBodyParts(8, 5, 5),
    strength: 10,
    movement: 3,
    actions: 9,
    aim: 8,
  },
}
export type CharacterStatsType = keyof typeof charactersStatsDefinition
export const characterStats: { [key in CharacterStatsType]: CharacterStats } = charactersStatsDefinition

const characterCreatorsDefinition = {
  player: (world: TlbWorld) => createPlayer(world),
  eliteGuard: (world: TlbWorld) => {
    const entity = createDefaultNpc(world, 'some elite guard', 'eliteGuard', 'eliteGuard')
    addDialog(world, entity, 'guardRandomRemarks')
    return entity
  },
  guard: (world: TlbWorld) => {
    const entity = createDefaultNpc(world, 'some guard', 'guard', 'guard')
    addDialog(world, entity, 'guardRandomRemarks')
    return entity
  },
}

export const defaultActions: ActionType[] = ['longMove', 'hit', 'rush', 'endTurn']

export function createPlayer(world: TlbWorld): Entity {
  const player = createCharacter(world, 'player', 'player', defaultActions)
  world
    .editEntity(player)
    .withComponent<{}>('player', {})
    .withComponent<FovComponent>('fov', { fov: [] })
    .withComponent<{}>('viewport-focus', {})

  take(world, player, 'idCard')
  take(world, player, 'rifle')
  take(world, player, 'sniperRifle')
  take(world, player, 'bootsOfStriding')
  equip(world, player, 'nailGun', ['leftArm'])
  equip(world, player, 'leatherJacket', ['torso', 'leftArm', 'rightArm'])
  return player
}

export function createDefaultNpc(world: TlbWorld, name: string, statsType: CharacterStatsType, featureType: FeatureType): Entity {
  const npc = createEmptyNpc(world, name, statsType, featureType, defaultActions)
  equip(world, npc, 'rifle', ['leftArm'])
  return npc
}

export function take(world: TlbWorld, entity: Entity, type: ItemType): Entity {
  const item = world.createEntity().withComponent<ItemComponent>('item', { type }).entity
  world.getComponent<InventoryComponent>(entity, 'inventory')!.content.push(item)
  return item
}

export function equip(world: TlbWorld, entity: Entity, type: ItemType, bodyParts: string[]): void {
  const item = take(world, entity, type)
  world.getComponent<EquipedItemsComponent>(entity, 'equiped-items')!.equipment.push({
    entity: item,
    bodyParts,
  })
}

export function createEmptyNpc(
  world: TlbWorld,
  name: string,
  statsType: CharacterStatsType,
  featureType: FeatureType,
  actions: ActionType[]
): Entity {
  const character = createCharacter(world, statsType, featureType, actions)
  return world
    .editEntity(character)
    .withComponent('npc', {})
    .withComponent<TriggersComponent>('triggers', { entities: [], type: 'dialog', name })
    .withComponent<AiComponent>('ai', { type: 'rushing', state: 'idle', interest: undefined, distrust: 0 }).entity
}

export function createCharacter(world: TlbWorld, statsType: CharacterStatsType, featureType: FeatureType, actions: ActionType[]): Entity {
  return world
    .createEntity()
    .withComponent<HasActionComponent>('has-action', { actions })
    .withComponent<InventoryComponent>('inventory', { content: [] })
    .withComponent<EquipedItemsComponent>('equiped-items', { equipment: [] })
    .withComponent<FeatureComponent>('feature', { feature: () => features[featureType] })
    .withComponent<CharacterStatsComponent>('character-stats', createCharacterStatsComponent(statsType))
    .withComponent<ActiveEffectsComponent>('active-effects', { effects: [] }).entity
}

export type CharacterCreatorType = keyof typeof characterCreatorsDefinition
export const characterCreators: { [key in CharacterCreatorType]: CharacterCreator } = characterCreatorsDefinition
