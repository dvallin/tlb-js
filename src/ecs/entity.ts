import { World } from "./world"

export type Entity = number

export class EntityModifier<C, S, R> {

    public constructor(
        private readonly world: World<C, S, R>,
        public readonly entity: Entity
    ) { }

    public withComponent<T>(name: C, component: T): EntityModifier<C, S, R> {
        this.world.getStorage(name).insert(this.entity, component)
        return this
    }

    public removeComponent(name: C): EntityModifier<C, S, R> {
        this.world.getStorage(name).remove(this.entity)
        return this
    }

    public delete(): void {
        this.world.components.forEach(s => s.remove(this.entity))
    }
}
