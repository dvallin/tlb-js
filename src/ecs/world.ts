import { Storage } from './storage'
import { System } from './system'
import { Resource } from './resource'
import { Entity, EntityModifier } from './entity'

export class World<C, S, R> {
  public readonly components: Map<C, Storage<object>> = new Map()
  public readonly resources: Map<R, Resource<C, S, R>> = new Map()
  public readonly systems: Map<S, System<C, S, R>> = new Map()
  public readonly emptySystems: Set<S> = new Set()
  public readonly activeSystems: Set<S> = new Set()

  private openEntities: Set<Entity> = new Set()
  private lastEntity: Entity = -1

  public get entityCount(): number {
    return this.lastEntity + 1 - this.openEntities.size
  }

  public registerComponentStorage<T extends object>(component: C, storage: Storage<T>): void {
    this.components.set(component, storage)
  }

  public getStorage<T extends object>(component: C): Storage<T> {
    return this.components.get(component)! as Storage<T>
  }

  public getResource<T extends Resource<C, S, R>>(resource: R): T {
    return this.resources.get(resource)! as T
  }

  public getComponent<T extends object>(entity: Entity, component: C): T | undefined {
    const storage = this.getStorage<T>(component)
    if (storage !== undefined) {
      return storage.get(entity)
    }
    return undefined
  }

  public hasEntity(entity: Entity): boolean {
    return entity <= this.lastEntity && !this.openEntities.has(entity)
  }

  public hasComponent(entity: Entity, component: C): boolean {
    const storage = this.getStorage(component)
    if (storage !== undefined) {
      return storage.has(entity)
    }
    return false
  }

  public createEntity(): EntityModifier<C, S, R> {
    let entity
    if (this.openEntities.size > 0) {
      entity = this.openEntities.values().next().value
      this.openEntities.delete(entity)
    } else {
      entity = ++this.lastEntity
    }
    return new EntityModifier(this, entity)
  }

  public editEntity(entity: Entity): EntityModifier<C, S, R> {
    return new EntityModifier(this, entity)
  }

  public deleteEntity(entity: Entity): void {
    this.editEntity(entity).delete()
    this.openEntities.add(entity)
  }

  public registerSystem(name: S, system: System<C, S, R>): void {
    this.systems.set(name, system)
  }

  public enableSystem(name: S): void {
    this.activeSystems.add(name)
  }

  public disableSystem(name: S): void {
    this.activeSystems.delete(name)
  }

  public activeSystemsList(): S[] {
    const l: S[] = []
    this.activeSystems.forEach(s => l.push(s))
    return l
  }

  public registerResource(resource: Resource<C, S, R>): void {
    this.resources.set(resource.kind, resource)
  }

  public updateResources(): void {
    for (const resource of this.resources) {
      resource[1].update(this)
    }
  }

  public updateSystems(): void {
    this.emptySystems.clear()
    for (const name of this.activeSystems) {
      const system = this.systems.get(name)!
      const firstStorage = this.getStorage(system.components[0])
      let calls = 0
      firstStorage.foreach((entity: Entity) => {
        let callSystem = true
        for (let i = 1; i < system.components.length; i++) {
          if (!this.hasComponent(entity, system.components[i])) {
            callSystem = false
            break
          }
        }
        if (callSystem) {
          system.update(this, entity)
          calls++
        }
      })
      if (calls === 0) {
        this.emptySystems.add(name)
      }
    }
  }
}
