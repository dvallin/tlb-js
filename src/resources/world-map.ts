import { DiscreteSpace, Vector, Space } from '../spatial'
import { Entity } from '../ecs/entity'
import { ResourceName, TlbResource, TlbWorld } from '../tlb'
import { FeatureComponent, features } from '../components/feature'
import { Shape } from '../geometry/shape'
import { SetSpace, DiscreteSetSpace } from '../spatial/set-space'
import { FovComponent } from '../components/fov'
import { Rectangle } from '../geometry/rectangle'

export class WorldMap implements TlbResource {
  public readonly kind: ResourceName = 'map'

  public readonly tiles: Space<Entity> = new DiscreteSpace<Entity>()
  public readonly characters: Space<Entity> = new DiscreteSpace<Entity>()

  public boundaries = new Rectangle(0, 0, 0, 0)

  public visible: SetSpace = new DiscreteSetSpace()
  public readonly discovered: SetSpace = new DiscreteSetSpace()

  public constructor() {}

  public update(world: TlbWorld): void {
    this.visible = new DiscreteSetSpace()
    world.components.get('player')!.foreach(player => {
      const fov = world.getComponent<FovComponent>(player, 'fov')
      if (fov !== undefined) {
        fov.fov.forEach(p => {
          this.visible.set(p)
          this.discovered.set(p)
        })
      }
    })
  }

  public setTile(position: Vector, entity: Entity): void {
    this.tiles.set(position, entity)
    this.boundaries = this.boundaries.cover(position)
  }
  public setCharacter(position: Vector, entity: Entity): void {
    this.characters.set(position, entity)
    this.boundaries = this.boundaries.cover(position)
  }
  public getTile(position: Vector): Entity | undefined {
    return this.tiles.get(position)
  }
  public getCharacter(position: Vector): Entity | undefined {
    return this.characters.get(position)
  }
  public removeTile(position: Vector): Entity | undefined {
    return this.tiles.remove(position)
  }
  public removeCharacter(position: Vector): Entity | undefined {
    return this.characters.remove(position)
  }

  public isTileBlocking(world: TlbWorld, position: Vector): boolean {
    return this.tileMatches(world, position, f => {
      if (f === undefined) {
        return true
      }
      return features[f.type].blocking
    })
  }

  public isDiscovered(position: Vector): boolean {
    return this.discovered.has(position)
  }

  public isLightBlocking(world: TlbWorld, position: Vector): boolean {
    return this.tileMatches(world, position, f => {
      if (f === undefined) {
        return true
      }
      return features[f.type].lightBlocking
    })
  }

  public tileMatches(world: TlbWorld, position: Vector, predicate: (f: FeatureComponent | undefined) => boolean): boolean {
    const entity = this.tiles.get(position)
    let feature: FeatureComponent | undefined
    if (entity !== undefined) {
      feature = world.getComponent<FeatureComponent>(entity, 'feature')
    }
    return predicate(feature)
  }

  public isShapeFree(world: TlbWorld, shape: Shape): boolean {
    return this.shapeHasAll(world, shape, f => f === undefined)
  }

  public isShapeBlocked(world: TlbWorld, shape: Shape): boolean {
    return this.shapeHasSome(world, shape, f => f !== undefined)
  }

  public shapeHasAll(world: TlbWorld, shape: Shape, predicate: (f: FeatureComponent | undefined) => boolean): boolean {
    return shape.all(p => this.tileMatches(world, p, predicate))
  }

  public shapeHasSome(world: TlbWorld, shape: Shape, predicate: (f: FeatureComponent | undefined) => boolean): boolean {
    return shape.some(p => this.tileMatches(world, p, predicate))
  }
}
