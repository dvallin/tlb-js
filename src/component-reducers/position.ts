import { WorldMapResource } from '../resources/world-map'
import { Vector } from '../spatial'
import { TlbWorld } from '../tlb'
import { Entity } from '../ecs/entity'
import { PositionComponent } from '../components/position'

export function move(world: TlbWorld, entity: Entity, level: number, position: Vector): void {
  const l = world.getResource<WorldMapResource>('map').levels[level]
  if (l !== undefined && !l.isBlocking(world, position, entity)) {
    const oldPosition = world.getComponent<PositionComponent>(entity, 'position')!
    l.removeCharacter(oldPosition.position)
    l.setCharacter(position, entity)
    world.editEntity(entity).withComponent<PositionComponent>('position', { level, position })
  }
}
