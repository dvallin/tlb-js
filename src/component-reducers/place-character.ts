import { Entity } from '../ecs/entity'
import { PositionComponent, centeredPosition } from '../components/position'
import { TlbWorld } from '../tlb'
import { WorldMapResource } from '../resources/world-map'
import { Vector } from '../spatial'

export function placeCharacter(world: TlbWorld, entity: Entity, level: number, position: Vector): void {
  const map = world.getResource<WorldMapResource>('map')
  map.levels[level].setCharacter(position, entity)
  world.editEntity(entity).withComponent<PositionComponent>('position', centeredPosition(level, position))
}
