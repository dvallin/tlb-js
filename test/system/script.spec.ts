import { Script } from '../../src/systems/script'
import { registerComponents, TlbWorld } from '../../src/tlb'
import { World } from '../../src/ecs/world'
import { ScriptComponent } from '../../src/components/script'
import { Vector } from '../../src/spatial'
import { characterCreators } from '../../src/assets/characters'
import { placeCharacter } from '../../src/component-reducers/place-character'
import { WorldMapResource } from '../../src/resources/world-map'
import { Entity } from '../../src/ecs/entity'
import { PositionComponent } from '../../src/components/position'

describe('Npc', () => {
  let world: TlbWorld
  let system: Script
  let player: Entity
  beforeEach(() => {
    world = new World()
    registerComponents(world)
    world.registerResource(new WorldMapResource(4))

    player = characterCreators.player(world)
    placeCharacter(world, player, 0, new Vector([0, 0]))

    system = new Script()
  })

  it('removes script on empty path', () => {
    world.editEntity(player).withComponent<ScriptComponent>('script', { path: [] }).entity
    system.update(world, player)
    expect(world.hasComponent(player, 'script')).toBeFalsy()
  })

  it('moves character towards waypoint', () => {
    world.editEntity(player).withComponent<ScriptComponent>('script', { path: [new Vector([1, 0])] }).entity
    system.update(world, player)
    expect(world.getComponent<PositionComponent>(player, 'position')!.position).toMatchSnapshot()
  })

  it('removes waypoint on already reached position', () => {
    world.editEntity(player).withComponent<ScriptComponent>('script', { path: [new Vector([0, 0])] }).entity
    system.update(world, player)
    expect(world.getComponent<ScriptComponent>(player, 'script')!.path).toHaveLength(0)
  })

  it('removes waypoint on overshooting position', () => {
    world.editEntity(player).withComponent<ScriptComponent>('script', { path: [new Vector([0.01, 0])] }).entity
    system.update(world, player)
    expect(world.getComponent<ScriptComponent>(player, 'script')!.path).toHaveLength(0)
  })
})
