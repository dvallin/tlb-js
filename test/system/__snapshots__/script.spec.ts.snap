// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Npc moves character towards waypoint 1`] = `
Vector {
  "coordinates": Array [
    0.9,
    0.25,
  ],
}
`;
